lib/rhc/commands/port_forward.rb

```diff
@@ -73,55 +74,77 @@ class PortForward < Base
     option ["-g", "--gear ID"], "Gear ID you are port forwarding to (optional)"
     option ["-s", "--service [SERVICE,]"], "A CSV list of services to port forward (optional)"
     def run(app)
-      rest_app = find_app
-      ssh_uri = URI.parse(options.gear ? rest_app.gear_ssh_url(options.gear) : rest_app.ssh_url)
-
-      say "Using #{ssh_uri}..." if options.debug
-
+      ssh_executable = check_ssh_executable! options.ssh
       forwarding_specs = []
 
       begin
-        say "Checking available ports ... "
-
-        Net::SSH.start(ssh_uri.host, ssh_uri.user) do |ssh|
-          # If a specific gear is targeted, do not include remote (e.g. database) ports
-          list_ports_cmd = "rhc-list-ports#{options.gear ? ' --exclude-remote' : ''}"
-          ssh.exec! list_ports_cmd do |channel, stream, data|
-            if stream == :stderr
-              data.each_line do |line|
-                line.chomp!
-                # FIXME: This is really brittle; there must be a better way
-                # for the server to tell us that permission (what permission?)
-                # is denied.
-                raise RHC::PermissionDeniedException.new "Permission denied." if line =~ /permission denied/i
-                # ...and also which services are available for the application
-                # for us to forward ports for.
-                if line =~ /\A\s*(\S+) -> #{HOST_AND_PORT}\z/ and (options.service.nil? or options.service.empty? or options.service.split(',').include? $1)
-                  debug fs = ForwardingSpec.new($1, $2, $3.to_i)
-                  forwarding_specs << fs
-                else
-                  debug line
-                end
+        rest_app = find_app
+        ssh_uri = URI.parse(options.gear ? rest_app.gear_ssh_url(options.gear) : rest_app.ssh_url)
+        debug "Using #{ssh_uri}..."
 
+        output = ""
+        say "Checking available ports ... "
+        # Only use Net::SSH if no ssh executable is specified
+        if !options.ssh
+          Net::SSH.start(ssh_uri.host, ssh_uri.user) do |ssh|
+            # If a specific gear is targeted, do not include remote (e.g. database) ports
+            list_ports_cmd = "rhc-list-ports#{options.gear ? ' --exclude-remote' : ''}"
+            ssh.exec! list_ports_cmd do |channel, stream, data|
+              if stream == :stderr
+                output = data
```

Comment:
Adding some debug statements, I can see that this block only gets called once. Here is some example output:
```
DEBUG: Using ssh://564100f01cbe7428cf000001@scaletest-admin.dev.rhcloud.com...
Checking available ports ... 
DEBUG: DATA FOUND:
haproxy -> 127.1.244.130:8080
haproxy -> 127.1.244.131:8080
httpd -> 127.1.244.129:8080
ruby -> 127.1.244.129:20332
DEBUG: COMPLETE OUTPUT:
haproxy -> 127.1.244.130:8080
haproxy -> 127.1.244.131:8080
httpd -> 127.1.244.129:8080
ruby -> 127.1.244.129:20332
```

And the change I made:
```
        if !options.ssh
          Net::SSH.start(ssh_uri.host, ssh_uri.user) do |ssh|
            # If a specific gear is targeted, do not include remote (e.g. database) ports
            ssh.exec! list_ports_cmd do |channel, stream, data|
              if stream == :stderr
+               debug "\nDATA FOUND:\n" + data
                output = data
              end
            end
          end
        else
          debug "Using #{ssh_executable} to determine forwarding ports."
          ssh_cmd = "#{ssh_executable} #{ssh_uri.user}@#{ssh_uri.host} '#{list_ports_cmd} 2>&1'"
          status, output = exec(ssh_cmd)
          if status != 0
            raise RHC::SSHCommandFailed.new(status, output)
          end
        end
+       debug "COMPLETE OUTPUT:\n" + output
```

However, I can definitely see your point. I don't think this block will ever be called more than once, but will make a change just in case.